format-avro/src/main/java/io/cdap/plugin/format/avro/output/DelegatingAvroKeyRecordWriter.java

```
@@ -0,0 +1,95 @@
+  private final CodecFactory codecFactory;
+  private final Function<TaskAttemptContext, OutputStream> outputStreamSupplier;
+  private final int syncInterval;
+
+  public DelegatingAvroKeyRecordWriter(TaskAttemptContext context,
+                                       CodecFactory codecFactory,
+                                       Function<TaskAttemptContext, OutputStream> outputStreamSupplier,
+                                       int syncInterval) {
+    super();
+    this.delegateMap = new HashMap<>();
+    this.context = context;
+    this.codecFactory = codecFactory;
+    this.outputStreamSupplier = outputStreamSupplier;
+    this.syncInterval = syncInterval;
+  }
+
+  @Override
+  public void write(AvroKey<GenericRecord> key, NullWritable value) throws IOException, InterruptedException {
+    RecordWriter<AvroKey<GenericRecord>, NullWritable> delegate =
+      delegateMap.computeIfAbsent(key.datum().getSchema().hashCode(), (k) -> {
+        try {
+          Configuration conf = context.getConfiguration();
+          GenericData dataModel = AvroSerialization.createDataModel(conf);
+
+          return new AvroKeyRecordWriter<>(key.datum().getSchema(),
+                                           dataModel,
+                                           codecFactory,
+                                           outputStreamSupplier.apply(context),
+                                           syncInterval);
+        } catch (IOException ioe) {
+          throw new RuntimeException("Unable to initialize Avro Record Writer", ioe);
+        }
+      });
+
+    delegate.write(key, value);
+  }
+
+  @Override
+  public void close(TaskAttemptContext context)  {
+    RuntimeException ex = new RuntimeException("Unable to close delegate.");
+
+    for (RecordWriter<AvroKey<GenericRecord>, NullWritable> delegate : delegateMap.values()) {
+      try {
+        delegate.close(context);
+      } catch (IOException | InterruptedException e) {
```

don't need to catch the InterruptedException, better to stop doing all work and let it propagate up immediately